src/main/kotlin/RabbitMQBroker.kt

```diff
@@ -0,0 +1,117 @@
+import com.rabbitmq.client.*
+import org.slf4j.LoggerFactory
+
+class RabbitMQBroker(uri: String = "amqp://guest:guest@localhost"): IMessageBroker {
+    private val queues = mutableMapOf<String, RabbitMQQueue>()
+    var connection: Connection
+    var channel: Channel
+    val exchangeName = "task-delayed-exchange"
+
+    private var logger = LoggerFactory.getLogger(this::class.java)
+
+    init {
+        val factory = ConnectionFactory()
+
+        if (false && logger.isDebugEnabled) {
+            factory.setTrafficListener(object : TrafficListener {
+                override fun write(outboundCommand: Command?) {
+                    logger.debug("RabbitMQ write: $outboundCommand")
+
+                }
+
+                override fun read(inboundCommand: Command?) {
+                    logger.debug("RabbitMQ read: $inboundCommand")
+                }
+            })
+        }
+
+        factory.setUri(uri)
+        connection = factory.newConnection()
+        channel = connection.createChannel()
+        channel.exchangeDeclare(
+            exchangeName,
+            "x-delayed-message",
+            true,
+            false,
+            mapOf("x-delayed-type" to "direct")
+        )
+
+    }
+
+    private fun getQueueById(queueName: QueueName): RabbitMQQueue {
+        return queues.getOrPut(queueName) { RabbitMQQueue(queueName, this) }
+    }
+
+    override fun submitMessage(queueName: QueueName, message: Message) {
+        getQueueById(queueName).submitMessage(message)
+    }
+
+    override fun startConsumer(queueName: QueueName, handler: ConsumerHandler): IConsumer {
+        return getQueueById(queueName).startConsumer(handler)
+    }
+
+    override fun close() {
+        connection.close()
+    }
+}
+const val HEADERS_PREFIX = "kot-"
+
+
+class RabbitMQQueue(val queueName: QueueName, val broker: RabbitMQBroker) {
+    private val pubChannel: Channel
+    init {
+        //pubChannel = broker.connection.createChannel()
+        pubChannel = broker.channel
+        pubChannel.queueDeclare(queueName, true, false, false, null)
+        pubChannel.queueBind(queueName, broker.exchangeName, queueName)
+    }
```

Comment:
just one more spoon of sugar (I love sugar)
```kotlin
private val pubChannel: Channel = broker.channel.apply {
       queueDeclare(queueName, true, false, false, null)
       queueBind(queueName, broker.exchangeName, queueName)
}
```

Reply:
OK. But I still believe that this kind of sugar moves software engineering focus from business to itself.